src/App.php

```diff
@@ -97,23 +98,74 @@ public function run($request = null, $response = null)
             $event->setResponse($response);
         }
 
-        $dispatcher = $this->getDispatcher();
-        $eventManager = $this->getEventManager();
+        return $event;
+    }
 
-        try {
-            $routeInfo = call_user_func($dispatcher, $event->getRequest());
+    /**
+     * Application execution.
+     *
+     * @param mixed|null $request  Representation of an outgoing,
+     *  client-side request.
+     * @param mixed|null $response Representation of an incoming,
+     *  server-side response.
+     *
+     * @return mixed
+     */
+    public function run($request = null, $response = null)
+    {
+        $event = $this->setUpEventWithRequestResponse($request, $response);
 
-            if (!($routeInfo instanceof RouteInfoInterface)) {
-                throw new RuntimeException('Dispatch does not return RouteInfo object');
-            }
+        $dispatcher   = $this->getDispatcher();
+        $eventManager = $this->getEventManager();
 
-            $event->setRouteInfo($routeInfo);
-            $event->setName($routeInfo->getName());
+        try {
+            $routeInfo = $dispatcher($event->getRequest());
+            $this->handleRoute($routeInfo, $dispatcher, $eventManager, $event);
         } catch (Exception $exception) {
             return $this->triggerWithException($eventManager, $event, 'dispatch_error', $exception)
                         ->getResponse();
         }
+        $this->handleResponse($eventManager, $event, $routeInfo);
 
+        return $event->getResponse();
+    }
+
+    /**
+     * Handle Route.
+     *
+     * @param RouteInfoInterface $routeInfo
+     * @param Dispatcher $dispatcher
+     * @param PennyEvmInterface $eventManager
+     * @param PennyEventInterface $event
+     *
+     * @throws RuntimeException if dispatch does not return RouteInfo object.
+     */
+    private function handleRoute(
+        RouteInfoInterface $routeInfo,
+        Dispatcher $dispatcher,
+        PennyEvmInterface $eventManager,
+        PennyEventInterface $event
+    ) {
+        if (!($routeInfo instanceof RouteInfoInterface)) {
```

Comment:
You can just write `if (!$routeInfo instanceof RouteInfoInterface) {` because the `instanceof` has higher operator precedence than `!`.